common/src/main/java/com/microsoft/identity/common/internal/ui/webview/certbasedauth/AbstractSmartcardCertBasedAuthChallengeHandler.java

```diff
@@ -88,57 +89,84 @@ public Void processChallenge(ClientCertRequest request) {
         mCbaManager.requestDeviceSession(new AbstractSmartcardCertBasedAuthManager.ISessionCallback() {
             @Override
             public void onGetSession(@NonNull final ISmartcardSession session) throws Exception {
-                if (session.getPinAttemptsRemaining() == 0) {
-                    promptTooManyFailedPinAttempts(methodTag);
-                    request.cancel();
-                    onPausedSmartcardDiscovery();
-                    return;
-                }
-                //Create List that contains cert details only pertinent to the cert picker.
+                final int pinAttemptsRemaining = session.getPinAttemptsRemaining();
                 final List<ICertDetails> certList = session.getCertDetailsList();
-                //If no certs were found, cancel flow.
-                if (certList.isEmpty()) {
-                    Logger.info(methodTag,  NO_PIV_CERTS_FOUND_MESSAGE);
-                    mTelemetryHelper.setResultFailure(NO_PIV_CERTS_FOUND_MESSAGE);
-                    mDialogHolder.showErrorDialog(
-                            R.string.smartcard_no_cert_dialog_title,
-                            R.string.smartcard_no_cert_dialog_message);
-                    request.cancel();
-                    onPausedSmartcardDiscovery();
-                    return;
-                }
+                pauseToCloseConnection(new IDisconnectionCallback() {
+                    @Override
+                    public void onClosedConnection() {
+                        if (pinAttemptsRemaining == 0) {
+                            promptTooManyFailedPinAttempts(methodTag);
+                            request.cancel();
+                            return;
+                        }
+                        //If no certs were found, cancel flow.
+                        if (certList.isEmpty()) {
+                            Logger.info(methodTag, NO_PIV_CERTS_FOUND_MESSAGE);
+                            mTelemetryHelper.setResultFailure(NO_PIV_CERTS_FOUND_MESSAGE);
+                            mDialogHolder.showErrorDialog(
+                                    R.string.smartcard_no_cert_dialog_title,
+                                    R.string.smartcard_no_cert_dialog_message);
+                            request.cancel();
+                            return;
+                        }
 
-                //Build and show Smartcard cert picker, which also handles the rest of the smartcard CBA flow.
-                mDialogHolder.showCertPickerDialog(
-                        certList,
-                        getSmartcardCertPickerDialogPositiveButtonListener(request),
-                        new SmartcardCertPickerDialog.CancelCbaCallback() {
-                            @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
+                        final ICancelCbaCallback cancelCallback = getGeneralCancelCbaCallback(request);
+                        mDialogHolder.showCertPickerDialog(
```

Comment:
I think some coments to explain why we are using showCertPickerDialog (I'm a little confused because we closed the connection already, right?)

Reply:
I think some terminology is confusing (and this could very well be from CBA documentation side and out of your control), for example I was misinterpreting "CloseConnection" to mean something relating to closing a network connection, but it can refer to physically removing a smart card from the device. Just something to think about if we can create some terminology that would be more specific to what is happening, if not maybe we can increase documentation to make it easier for developers to understand what a method is doing.

Reply:
`CloseConnection` refers to making sure the user removes their NFC smartcard before we proceed with the rest of the logic. For NFC, we want to only have the connection open when it is necessary (this is unlike USB), so I changed the logic to get the necessary data, then close the connection before proceeding with that data. So, before `pauseToCloseConnection` is called, we get the `pinAttemptsRemaining` and `certList`, then close the connection, since we don't need it to be open in order to show a certpicker dialog (using the certList).

I can add "smartcard" back into the method name, and maybe rename it to something like `pauseSmartcardConnection`, and maybe leave "close" out of it, if that's confusing. That would also make it a bit more applicable for the USB side (where the connection isn't being closed at all here, but we are pausing, in a sense, from reading off of the card).

Reply:
I think you can keep the existing method names, just adding comment would be helpful

Reply:
See lines 95-97, 115-116